qa/L0_response_cache/ensemble_cache_test.py

```diff
@@ -0,0 +1,182 @@
+
+import os
+import re
+import sys
+
+sys.path.append("../common")
+sys.path.append("../clients")
+
+import logging
+import unittest
+
+import infer_util as iu
+import numpy as np
+import test_util as tu
+import tritonclient.grpc as grpcclient
+
+
+class EnsembleCacheTest(tu.TestResultCollector):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.ensemble_model_name = "simple_graphdef_float32_float32_float32"
+        self.composing_model_name = "graphdef_float32_float32_float32"
+        self.model_directory = os.path.join(os.getcwd(), "models", "ensemble_models")
+        input0_data = np.ones((1, 16), dtype=np.float32)
+        input1_data = np.ones((1, 16), dtype=np.float32)
+        self.input_tensors = [
+            grpcclient.InferInput("INPUT0", [1, 16], "FP32"),
+            grpcclient.InferInput("INPUT1", [1, 16], "FP32"),
+        ]
+        self.input_tensors[0].set_data_from_numpy(input0_data)
+        self.input_tensors[1].set_data_from_numpy(input1_data)
+
+    def _get_infer_result(self):
+        triton_client = grpcclient.InferenceServerClient("localhost:8001", verbose=True)
+        results = triton_client.infer(
+            model_name=self.ensemble_model_name, inputs=self.input_tensors
+        )
+        stats = triton_client.get_inference_statistics(
+            model_name=self.ensemble_model_name, as_json=True
+        )
+        return stats
+
+    def _run_ensemble(self):
+        stats = self._get_infer_result()
+        stats = self._get_infer_result()
+        return stats["model_stats"][1]["inference_stats"]
+
+    def _update_config(self, config_file_path, config_pattern, config_to_add):
+        with open(config_file_path, "r") as f:
+            config_data = f.read()
+            if config_pattern not in config_data:
+                with open(config_file_path, "a") as f:
+                    f.write(config_to_add)
+
+    def _enable_response_cache_for_all_models(self):
+        contents = os.listdir(self.model_directory)
+        folders = [
+            folder
+            for folder in contents
+            if os.path.isdir(os.path.join(self.model_directory, folder))
+        ]
+        for model_dir in folders:
+            config_file_path = os.path.join(
+                self.model_directory, str(model_dir), "config.pbtxt"
+            )
+            config_pattern = "response_cache"
+            config_to_add = "response_cache {\n  enable:true\n}\n"
+            self._update_config(config_file_path, config_pattern, config_to_add)
+
+    def _remove_extra_config(self, config_file_path, config_to_remove):
+        with open(config_file_path, "r") as f:
+            config_data = f.read()
+        updated_config_data = re.sub(config_to_remove, "", config_data)
+        with open(config_file_path, "w") as f:
+            f.write(updated_config_data)
+
+    def test_ensemble_top_level_cache(self):
```

Comment:
Do you have any tests that validate the output results are actually correct? We should check somewhere that the cached outputs are correct.

Reply:
I will add a test.